### PR TITLE
Adding offset to web services for pagination or infinite scroll

### PR DESCRIPTION
**Pull request by ghost**
Questions | Answers
-- | --
Branch? | develop
Description? | You can find a new push into PrestaShop-webservice-lib for offset param, I will create request-pull for it. PrestaShop/PrestaShop-webservice-lib#59
Type? | improvement
Category? | WS
How to test? | view bellow<br><pre>define('DEBUG', false);<br>define('_PS_DEBUG_SQL_', false);<br>define('PS_SHOP_PATH', 'http://localhost/fortnight');<br>define('PS_WS_AUTH_KEY', 'ZKYA3SF94R1J6GWVREKKKCD3ZDRKZ36P'); // change by your ws_key code<br>require_once('PSWebServiceLibrary.php');<br><br>$webService = new PrestaShopWebservice(PS_SHOP_PATH, PS_WS_AUTH_KEY, DEBUG);<br>$opt['output_format'] = 'JSON';<br>$opt['ps_method']   = 'GET';<br>$output = $webService->get(array(<br>    'resource' => 'products',<br>    'display' => '[id,name]',<br>    'limit' => 5,<br>    'offset' => 5<br>));<br>$json = json_encode($output);<br>echo $json;</pre>

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/16381)

### CODE REVIEW

File: classes/webservice/WebserviceRequest.php [hunk 1]
```diff
@@ -366,8 +366,12 @@ public static function getResources()
         return $resources;
     }
 
-    /** @todo Check how get parameters */
-    /** @todo : set this method out */
+    /**
+     * @todo Check how get parameters
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
     * @todo Check how get parameters
```

File: classes/webservice/WebserviceRequest.php [hunk 2]
```diff
@@ -366,8 +366,12 @@ public static function getResources()
         return $resources;
     }
 
-    /** @todo Check how get parameters */
-    /** @todo : set this method out */
+    /**
+     * @todo Check how get parameters 
+     */
+    /**
+     * @todo : set this method out
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
     * @todo : set this method out
```

File: classes/webservice/WebserviceRequest.php [hunk 3]
```diff
@@ -388,13 +392,15 @@ public function getPriceForProduct($field, $entity_object, $ws_params)
...
+    /**
+     * @todo : set this method out 
+     */
 
     /**
      * This method is used for calculate the price for products on a virtual fields.
      *
      * @param $entity_object
-     * @param array $parameters
+     * @param array         $parameters
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
     * @param array $parameters
```

File: classes/webservice/WebserviceRequest.php [hunk 4]
```diff
@@ -450,13 +456,15 @@ public function specificPriceCalculation($parameters)
         return $arr_return;
     }
 
-    /** @todo : set this method out */
+    /**
+     * @todo : set this method out
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
     * @todo : set this method out
```

File: classes/webservice/WebserviceRequest.php [hunk 5]
```diff
@@ -450,13 +456,15 @@ public function specificPriceCalculation($parameters)
...
+    /**
+     * @todo : set this method out 
+     */
 
     /**
      * This method is used for calculate the price for products on a virtual fields.
      *
      * @param $entity_object
-     * @param array $parameters
+     * @param array         $parameters
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
     * @param array $parameters
```

File: classes/webservice/WebserviceRequest.php [hunk 6]
```diff
@@ -642,9 +652,9 @@ protected function webserviceChecks()
     /**
      * Set a webservice error.
      *
-     * @param int $status
+     * @param int    $status
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
     * @param int $status
```

File: classes/webservice/WebserviceRequest.php [hunk 7]
```diff
@@ -642,9 +652,9 @@ protected function webserviceChecks()
...
      * @param string $label
-     * @param int $code
+     * @param int    $code
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
     * @param int $code
```

File: classes/webservice/WebserviceRequest.php [hunk 8]
```diff
@@ -661,11 +671,11 @@ public function setError($status, $label, $code)
     /**
      * Set a webservice error and propose a new value near from the available values.
      *
-     * @param int $num
+     * @param int    $num
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
     * @param int $num
```

File: classes/webservice/WebserviceRequest.php [hunk 9]
```diff
@@ -661,11 +671,11 @@ public function setError($status, $label, $code)
...
      * @param string $label
-     * @param array $value
-     * @param array $values
-     * @param int $code
+     * @param array  $value
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
     * @param array $value
```

File: classes/webservice/WebserviceRequest.php [hunk 10]
```diff
@@ -661,11 +671,11 @@ public function setError($status, $label, $code)
...
+     * @param array  $values
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
     * @param array $values
```

File: classes/webservice/WebserviceRequest.php [hunk 11]
```diff
@@ -661,11 +671,11 @@ public function setError($status, $label, $code)
...
+     * @param int    $code
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
     * @param int $code
```

File: classes/webservice/WebserviceRequest.php [hunk 12]
```diff
@@ -676,7 +686,7 @@ public function setErrorDidYouMean($num, $label, $value, $available_values, $cod
      * Return the nearest value picked in the values list.
      *
      * @param string $input
-     * @param array $words
+     * @param array  $words
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
     * @param array $words
```

File: classes/webservice/WebserviceRequest.php [hunk 13]
```diff
@@ -703,10 +713,10 @@ protected function getClosest($input, $words)
     /**
      * Used to replace the default PHP error handler, in order to display PHP errors in a XML format.
      *
-     * @param string $errno contains the level of the error raised, as an integer
-     * @param array $errstr contains the error message, as a string
-     * @param array $errfile errfile, which contains the filename that the error was raised in, as a string
-     * @param array $errline errline, which contains the line number the error was raised at, as an integer
+     * @param string $errno   contains the level of the error raised, as an integer
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
     * @param string $errno contains the level of the error raised, as an integer
```

File: classes/webservice/WebserviceRequest.php [hunk 14]
```diff
@@ -703,10 +713,10 @@ protected function getClosest($input, $words)
...
+     * @param array  $errstr  contains the error message, as a string
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
     * @param array $errstr contains the error message, as a string
```

File: classes/webservice/WebserviceRequest.php [hunk 15]
```diff
@@ -703,10 +713,10 @@ protected function getClosest($input, $words)
...
+     * @param array  $errfile errfile, which contains the filename that the error was raised in, as a string
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
     * @param array $errfile errfile, which contains the filename that the error was raised in, as a string
```

File: classes/webservice/WebserviceRequest.php [hunk 16]
```diff
@@ -703,10 +713,10 @@ protected function getClosest($input, $words)
...
+     * @param array  $errline errline, which contains the line number the error was raised at, as an integer
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
     * @param array $errline errline, which contains the line number the error was raised at, as an integer
```

File: classes/webservice/WebserviceRequest.php [hunk 17]
```diff
@@ -1099,10 +1109,12 @@ protected function manageFilters()
...
         $i18n_available_filters = array();
         foreach ($this->resourceConfiguration['fields'] as $fieldName => $field) {
-            if ((!isset($this->resourceConfiguration['hidden_fields']) ||
-                (isset($this->resourceConfiguration['hidden_fields']) && !in_array($fieldName, $this->resourceConfiguration['hidden_fields'])))) {
-                if ((!isset($field['i18n']) ||
-                (isset($field['i18n']) && !$field['i18n']))) {
+            if ((!isset($this->resourceConfiguration['hidden_fields']) 
+                || (isset($this->resourceConfiguration['hidden_fields']) && !in_array($fieldName, $this->resourceConfiguration['hidden_fields'])))
+            ) {
+                if ((!isset($field['i18n'])
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
                if ((!isset($field['i18n'])
```

File: classes/webservice/WebserviceRequest.php [hunk 18]
```diff
@@ -1255,7 +1267,9 @@ protected function manageFilters()
                     }
                     $sql_sort .= 'main_i18n.`' . bqSQL($this->resourceConfiguration['fields'][$fieldName]['sqlId']) . '` ' . $direction . ', '; // ORDER BY main_i18n.`field` ASC|DESC
                 } else {
-                    /** @var ObjectModel $object */
+                    /**
+ * @var ObjectModel $object
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
                     * @var ObjectModel $object 
```

File: classes/webservice/WebserviceRequest.php [hunk 19]
```diff
@@ -1255,7 +1267,9 @@ protected function manageFilters()
                     }
                     $sql_sort .= 'main_i18n.`' . bqSQL($this->resourceConfiguration['fields'][$fieldName]['sqlId']) . '` ' . $direction . ', '; // ORDER BY main_i18n.`field` ASC|DESC
                 } else {
-                    /** @var ObjectModel $object */
+                    /**
+ * @var ObjectModel $object 
+*/
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
                    */
```

File: classes/webservice/WebserviceRequest.php [hunk 20]
```diff
@@ -1281,10 +1295,22 @@ protected function manageFilters()
...
         }
+
+        $sql_offset = '';
+        if (isset($this->urlFragments['offset'])) {
+            $offsetArgs = explode(',', $this->urlFragments['offset']);
+            if (count($offsetArgs) > 2) {
+                $this->setError(400, 'The "offset" value has to be formed as this example: 10"', 39);
+                return false;
+            } else {
+                $sql_offset .= ' OFFSET '.(int)($offsetArgs[0]).(isset($offsetArgs[1]) ? ', '.(int)($offsetArgs[1]) : '')."\n";//
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
                $sql_offset .= ' OFFSET ' . (int) ($offsetArgs[0]) . (isset($offsetArgs[1]) ? ', ' . (int) ($offsetArgs[1]) : '') . "\n"; 
```

File: classes/webservice/WebserviceRequest.php [hunk 21]
```diff
@@ -1501,7 +1527,9 @@ protected function saveEntityFromXml($successReturnCode)
             return;
         }
 
-        /** @var SimpleXMLElement|Countable $xmlEntities */
+        /**
+ * @var SimpleXMLElement|Countable $xmlEntities
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
         * @var SimpleXMLElement|Countable $xmlEntities 
```

File: classes/webservice/WebserviceRequest.php [hunk 22]
```diff
@@ -1501,7 +1527,9 @@ protected function saveEntityFromXml($successReturnCode)
             return;
         }
 
-        /** @var SimpleXMLElement|Countable $xmlEntities */
+        /**
+ * @var SimpleXMLElement|Countable $xmlEntities 
+*/
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
        */
```

File: classes/webservice/WebserviceRequest.php [hunk 23]
```diff
@@ -1533,7 +1561,9 @@ protected function saveEntityFromXml($successReturnCode)
 
         $postponeNTreeRegeneration = false;
         foreach ($xmlEntities as $xmlEntity) {
-            /** @var SimpleXMLElement $xmlEntity */
+            /**
+ * @var SimpleXMLElement $xmlEntity 
+*/
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
        */
```

File: classes/webservice/WebserviceRequest.php [hunk 24]
```diff
@@ -1533,7 +1561,9 @@ protected function saveEntityFromXml($successReturnCode)
 
         $postponeNTreeRegeneration = false;
         foreach ($xmlEntities as $xmlEntity) {
-            /** @var SimpleXMLElement $xmlEntity */
+            /**
+ * @var SimpleXMLElement $xmlEntity
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
         * @var SimpleXMLElement $xmlEntity 
```

File: classes/webservice/WebserviceRequest.php [hunk 25]
```diff
@@ -1624,13 +1654,17 @@ protected function saveEntityFromXml($successReturnCode)
                     if ($result) {
                         if (isset($attributes->associations)) {
                             foreach ($attributes->associations->children() as $association) {
-                                /** @var SimpleXMLElement $association */
+                                /**
+ * @var SimpleXMLElement $association 
+*/
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
                                */
```

File: classes/webservice/WebserviceRequest.php [hunk 26]
```diff
@@ -1624,13 +1654,17 @@ protected function saveEntityFromXml($successReturnCode)
                     if ($result) {
                         if (isset($attributes->associations)) {
                             foreach ($attributes->associations->children() as $association) {
-                                /** @var SimpleXMLElement $association */
+                                /**
+ * @var SimpleXMLElement $association
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
                                 * @var SimpleXMLElement $association 
```

File: classes/webservice/WebserviceRequest.php [hunk 27]
```diff
@@ -1624,13 +1654,17 @@ protected function saveEntityFromXml($successReturnCode)
...
+ * @var SimpleXMLElement $association 
+*/
                                 // associations
                                 if (isset($this->resourceConfiguration['associations'][$association->getName()])) {
                                     $assocItems = $association->children();
                                     $values = array();
                                     foreach ($assocItems as $assocItem) {
-                                        /** @var SimpleXMLElement $assocItem */
+                                        /**
+ * @var SimpleXMLElement $assocItem
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
                                        * @var SimpleXMLElement $assocItem 
```

File: classes/webservice/WebserviceRequest.php [hunk 28]
```diff
@@ -1624,13 +1654,17 @@ protected function saveEntityFromXml($successReturnCode)
...
+*/
                                 // associations
                                 if (isset($this->resourceConfiguration['associations'][$association->getName()])) {
                                     $assocItems = $association->children();
                                     $values = array();
                                     foreach ($assocItems as $assocItem) {
-                                        /** @var SimpleXMLElement $assocItem */
+                                        /**
+ * @var SimpleXMLElement $assocItem 
+*/
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
                                       */
```

File: classes/webservice/WebserviceRequest.php [hunk 29]
```diff
@@ -1686,7 +1720,7 @@ protected function saveEntityFromXml($successReturnCode)
      *
      * @param string $sqlId
      * @param string $filterValue
-     * @param string $tableAlias = 'main.'
+     * @param string $tableAlias  = 'main.'
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
     * @param string $tableAlias = 'main.'
```

**Comment by Progi1984** [comment]
Hi @HamzaElghanoui, thank you for your contribution. Firstly, could you fix Pretty CI, please ? Thanks

File: classes/webservice/WebserviceRequest.php [hunk 30]
```diff
@@ -267,16 +267,16 @@ protected function getOutputObject($type)
         }
         $this->outputFormat = $type;
         switch ($type) {
-            case 'JSON':
-                require_once __DIR__ . '/WebserviceOutputJSON.php';
-                $obj_render = new WebserviceOutputJSON();
+        case 'JSON':
+            include_once __DIR__ . '/WebserviceOutputJSON.php';
```
Comment (PierreRambaud) [inline classes/webservice/WebserviceRequest.php]: What is the reason of changing these lines?

Comment (HamzaElghanoui) [reply classes/webservice/WebserviceRequest.php]: it's just remove spaces and replace require_once by include_once This is done by PrettyCI I think
now I recovery change include by required
thanks

Comment (PierreRambaud) [reply classes/webservice/WebserviceRequest.php]: You shouldn't run prettyci directly, we are using a specific configuration for php-cs-fixer and these files are already checked by it.

`./vendor/bin/php-cs-fixer fix`

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
You need to revert all your "Lint" changes. And squash your commit.
Currently we can't review this Pr because of all spaces you change.

The command to run cs fixer is `./vendor/bin/php-cs-fixer fix`

File: classes/webservice/WebserviceRequest.php (hunk 30 above)
Comment (HamzaElghanoui) [reply classes/webservice/WebserviceRequest.php]: ok there, I will
thanks a lot

**Comment by HamzaElghanoui** [comment]
./vendor/bin/php-cs-fixer fix
work for me but github PrettyCI -> Code formatting not valide How to fix this
thanks

File: classes/webservice/.php_cs.cache [hunk 31]
```diff
@@ -0,0 +1 @@
+{"php":"7.2.24-0ubuntu0.18.04.1","version":"2.15.1:v2.15.1#20064511ab796593a3990669eff5f5b535001f7c","indent":"    ","lineEnding":"\n","rules":{"blank_line_after_namespace":true,"braces":true,"class_definition":true,"elseif":true,"function_declaration":true,"indentation_type":true,"line_ending":true,"lowercase_constants":true,"lowercase_keywords":true,"method_argument_space":{"on_multiline":"ensure_fully_multiline"},"no_break_comment":true,"no_closing_tag":true,"no_spaces_after_function_name":true,"no_spaces_inside_parenthesis":true,"no_trailing_whitespace":true,"no_trailing_whitespace_in_comment":true,"single_blank_line_at_eof":true,"single_class_element_per_statement":{"elements":["property"]},"single_import_per_statement":true,"single_line_after_imports":true,"switch_case_semicolon_to_colon":true,"switch_case_space":true,"visibility_required":true,"encoding":true,"full_opening_tag":true},"hashes":{"WebserviceRequest.php":3642012977}}
```
Comment (PierreRambaud) [inline classes/webservice/.php_cs.cache]: Must be deleted

**Comment by PierreRambaud** [comment]
@HamzaElghanoui Weird, since php-cs-fixer use `.pĥp_cs.dist` :thinking: 
Could you please squash your commits?

File: classes/webservice/.php_cs.cache (hunk 31 above)
Comment (HamzaElghanoui) [reply classes/webservice/.php_cs.cache]: ok, dear, The file has been deleted when I tried to squash my commits just worked for last 4 commits What is required of me

**Pull request closed without merging**